win32ss/gdi/ntgdi/freetype.c

```diff
@@ -945,6 +1055,155 @@ SubstituteFontRecurse(LOGFONTW* pLogFont)
+    }
+}
+
+static NTSTATUS
+FontLink_Chain_LoadReg(PFONTLINK_CHAIN pChain, LPLOGFONTW pLF)
+{
+    NTSTATUS Status;
+    HKEY hKey;
+    DWORD cbData;
+    static const WCHAR s_szzDefFontLink[MAX_FONTLINK] =
+        L"tahoma.ttf,Tahoma\0"
+        L"msgothic.ttc,MS UI Gothic\0"
+        L"mingliu.ttc,PMingLiU\0"
+        L"simsun.ttc,SimSun\0"
+        L"gulim.ttc,Gulim\0"
+        L"\0";
+    static const WCHAR s_szzDefFixedPitchFontLink[MAX_FONTLINK] =
+        L"cour.ttf,Courier New\0"
+        L"msgothic.ttc,MS Gothic\0"
+        L"mingliu.ttc,MingLiU\0"
+        L"simsun.ttc,NSimSun\0"
+        L"gulim.ttc,GulimChe\0"
+        L"\0";
+
+    pChain->szzFontLink[0] = UNICODE_NULL;
+
+    // Open registry key
+    Status = RegOpenKey(
+        L"\\Registry\\Machine\\Software\\Microsoft\\Windows NT\\CurrentVersion\\FontLink\\SystemLink",
+        &hKey);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    // Load the FontLink entry
+    cbData = sizeof(pChain->szzFontLink);
+    Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, pChain->szzFontLink, &cbData);
+    if (!NT_SUCCESS(Status) && SubstituteFontRecurse(pLF)) // Retry with substituted name
+    {
+        cbData = sizeof(pChain->szzFontLink);
+        Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, pChain->szzFontLink, &cbData);
+    }
+
+    ZwClose(hKey); // Close the registry key
+
+    ASSERT(sizeof(pChain->szzFontLink) == sizeof(s_szzDefFontLink));
+
+    if (!NT_SUCCESS(Status))
+    {
+        if (!(pLF->lfPitchAndFamily & FIXED_PITCH))
+            RtlCopyMemory(pChain->szzFontLink, s_szzDefFontLink, sizeof(pChain->szzFontLink));
+        else
+            RtlCopyMemory(pChain->szzFontLink, s_szzDefFixedPitchFontLink, sizeof(pChain->szzFontLink));
+    }
+
+    // Avoid buffer overrun
+    pChain->szzFontLink[_countof(pChain->szzFontLink) - 1] = UNICODE_NULL;
+    pChain->szzFontLink[_countof(pChain->szzFontLink) - 2] = UNICODE_NULL;
+
+    return Status;
+}
+
+static NTSTATUS
+FontLink_Chain_Populate(PFONTLINK_CHAIN pChain)
+{
+    NTSTATUS Status;
+    SIZE_T cchLink;
+    PFONTLINK pLink;
+    LOGFONTW lfBase;
+    PTEXTOBJ pTextObj = pChain->pBaseTextObj;
+
+    InitializeListHead(&pChain->FontLinkList);
+
+    if (!pTextObj->TextFace[0])
+        return STATUS_INVALID_PARAMETER;
+
+    lfBase = pTextObj->logfont.elfEnumLogfontEx.elfLogFont;
+
+    // lfBase.lfFaceName <-- pTextObj->TextFace (if lfFaceName was empty)
+    if (!lfBase.lfFaceName[0])
+        RtlStringCchCopyW(lfBase.lfFaceName, _countof(lfBase.lfFaceName), pTextObj->TextFace);
+
+    // Load FontLink entry from registry
+    Status = FontLink_Chain_LoadReg(pChain, &lfBase);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    InitializeListHead(&pChain->FontLinkList);
+    LPWSTR pszzFontLink = pChain->szzFontLink;
+    LPWSTR pchLink = pszzFontLink;
+    while (*pchLink)
+    {
+        DPRINT("pchLink: '%S'\n", pchLink);
+        cchLink = wcslen(pchLink);
```

Comment:
Are the two function calls below modify pchLink such that you have to do the wcslen call here and not later?
Otherwise the `pchLink = &pchLink[cchLink + 1];` below could be modified into: `pchLink += wcslen(pchLink) + 1;`

Reply:
Fixed.